spring-kafka/src/test/java/org/springframework/kafka/listener/DeliveryAttemptAwareRetryListenerIntegrationTests.java

```diff
@@ -0,0 +1,308 @@
+					StringSerializer.class);
+			configProps.put(
+					ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG,
+					StringSerializer.class);
+			return new DefaultKafkaProducerFactory<>(configProps);
+		}
+
+		@Bean("customKafkaTemplate")
+		KafkaTemplate<String, String> kafkaTemplate() {
+			return new KafkaTemplate<>(producerFactory());
+		}
+	}
+
+	@EnableKafka
+	@Configuration
+	static class KafkaConsumerConfig {
+
+		@Autowired
+		EmbeddedKafkaBroker broker;
+
+		@Bean
+		KafkaAdmin kafkaAdmin() {
+			Map<String, Object> configs = new HashMap<>();
+			configs.put(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, this.broker.getBrokersAsString());
+			return new KafkaAdmin(configs);
+		}
+
+		@Bean
+		ConsumerFactory<String, String> consumerFactory() {
+			Map<String, Object> props = new HashMap<>();
+			props.put(
+					ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG,
+					this.broker.getBrokersAsString());
+			props.put(
+					ConsumerConfig.GROUP_ID_CONFIG,
+					"groupId");
+			props.put(
+					ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,
+					StringDeserializer.class);
+			props.put(
+					ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
+					StringDeserializer.class);
+			props.put(
+					ConsumerConfig.ALLOW_AUTO_CREATE_TOPICS_CONFIG, false);
+			props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+
+			return new DefaultKafkaConsumerFactory<>(props);
+		}
+
+		@Bean
+		ConcurrentKafkaListenerContainerFactory<String, String>
+		deliveryMyTestKafkaListenerContainerFactory0(ConsumerFactory<String, String> consumerFactory) {
+
+			final FixedBackOff fixedBackOff = new FixedBackOff(1, MAX_ATTEMPT_COUNT0);
+			DefaultErrorHandler errorHandler = new DefaultErrorHandler(fixedBackOff);
+			errorHandler.setRetryListeners(new DeliveryAttemptAwareRetryListener());
+
+			ConcurrentKafkaListenerContainerFactory<String, String> factory = new ConcurrentKafkaListenerContainerFactory<>();
+			factory.setConsumerFactory(consumerFactory);
+			factory.setCommonErrorHandler(errorHandler);
+
+
+			final ContainerProperties containerProperties = factory.getContainerProperties();
+			containerProperties.setDeliveryAttemptHeader(true);
+
+			return factory;
+		}
+
+		@Bean
+		ConcurrentKafkaListenerContainerFactory<String, String>
+		deliveryMyTestKafkaListenerContainerFactory1(ConsumerFactory<String, String> consumerFactory) {
+
+			final FixedBackOff fixedBackOff = new FixedBackOff(1, MAX_ATTEMPT_COUNT1);
+			DefaultErrorHandler errorHandler = new DefaultErrorHandler(fixedBackOff);
+			errorHandler.setRetryListeners(new DeliveryAttemptAwareRetryListener());
+
+			ConcurrentKafkaListenerContainerFactory<String, String> factory = new ConcurrentKafkaListenerContainerFactory<>();
+			factory.setConsumerFactory(consumerFactory);
+			factory.setCommonErrorHandler(errorHandler);
+
+
+			final ContainerProperties containerProperties = factory.getContainerProperties();
+			containerProperties.setDeliveryAttemptHeader(true);
+
+			return factory;
```

Comment:
Apparently a common code for both these two bean could be extracted to `private` method for better readability.